src/software/new_geom/vector.h

```diff
@@ -0,0 +1,521 @@
+ *
+ * @return the new value of Vector u
+ */
+Vector &operator-=(Vector &u, const Vector &v);
+
+/**
+ * Multiplies a vector by a scalar
+ *
+ * @param s the scaling factor
+ * @param v the vector to scale
+ *
+ * @return the scaled vector
+ */
+constexpr Vector operator*(double s, const Vector &v);
+
+/**
+ * Multiplies a vector by a scalar
+ *
+ * @param p the vector to scale
+ * @param s the scaling factor
+ *
+ * @return the scaled vector
+ */
+constexpr Vector operator*(const Vector &p, double s);
+
+/**
+ * Scales a vector by a scalar
+ *
+ * @param p the vector to scale
+ * @param s the scaling factor
+ *
+ * @return p scaled by the scaling factor
+ */
+Vector &operator*=(Vector &p, double s);
+
+/**
+ * Divides a vector by a scalar
+ *
+ * @param p the vector to scale
+ * @param s the scalar to divide by
+ *
+ * @return the scaled vector
+ */
+constexpr Vector operator/(const Vector &p, double s);
+
+/**
+ * Scales a vector by a scalar
+ *
+ * @param p the vector to scale
+ * @param s the sclaing factor
+ *
+ * @return p scaled by the scaling factor
+ */
+Vector &operator/=(Vector &p, double s);
+
+/**
+ * Prints a vector to a stream
+ *
+ * @param os the stream to print to
+ * @param v the Vector to print
+ *
+ * @return the stream with the Vector printed
+ */
+inline std::ostream &operator<<(std::ostream &os, const Vector &v);
+
+/**
+ * Compares two Vectors for equality
+ *
+ * @param u the first Vector
+ * @param v the second Vector
+ *
+ * @return true if the two vectors represent the same vector, and false otherwise
+ */
+constexpr bool operator==(const Vector &u, const Vector &v);
+
+/**
+ * Compares two vectors for inequality
+ *
+ * @param u the first Vector
+ * @param v the second Vector
+ *
+ * @return true if the two vectors represent different vectors, and false otherwise
+ */
+constexpr bool operator!=(const Vector &u, const Vector &v);
+
+inline Vector Vector::createFromAngle(const Angle &angle)
+{
+    return Vector(angle.cos(), angle.sin());
+}
+
+inline constexpr Vector::Vector() : _x(0.0), _y(0.0) {}
+
+inline constexpr Vector::Vector(double x, double y) : _x(x), _y(y) {}
+
+inline constexpr Vector::Vector(const Vector &v) : _x(v.x()), _y(v.y()) {}
+
+inline constexpr double Vector::x() const
+{
+    return _x;
+}
+
+inline constexpr double Vector::y() const
+{
+    return _y;
+}
+
+inline void Vector::set(double x, double y)
+{
+    this->_x = x;
+    this->_y = y;
+}
+
+inline void Vector::setX(double x)
+{
+    this->_x = x;
+}
+
+inline void Vector::setY(double y)
+{
+    this->_y = y;
+}
+
+inline constexpr double Vector::lensq() const
+{
+    return _x * _x + _y * _y;
+}
+
+inline double Vector::len() const
+{
+    return std::hypot(_x, _y);
+}
+
+inline Vector Vector::norm() const
+{
+    return len() < 1.0e-9 ? Vector() : Vector(_x / len(), _y / len());
+}
+
+inline Vector Vector::norm(double length) const
+{
+    return len() < 1.0e-9 ? Vector() : Vector(_x * length / len(), _y * length / len());
```

Comment:
Why are we rouding down like this?

Reply:
Actually the rounding somewhat makes sense, but is there a specific reason we're using `1.0e-9` here? I would imagine we'd want to compare to `EPSILON` instead (ie. check that the number is basically 0 in terms of what the computer is capable of representing).

Reply:
`1e-9` was the value of `NEAR`.... :roll_eyes:

Reply:
Haha, I think this is a good opportunity to question some of the decisions that were made when designing this originally. If we can't come up with a good reason for this value, I'd be inclined to use `EPSILON` (because that indicates 0 to within machine precision).